File: StompCoin/Managers/CoinManager.swift
import Foundation
import SwiftData

class CoinManager {
    static let shared = CoinManager()
    
    private init() {}
    
    // Coin-Generierung basierend auf Schritten
    func generateCoins(from steps: Int, user: User, modelContext: ModelContext) -> Double {
        let baseRate = 0.1 // 0.1 StompCoins pro Schritt
        var multiplier = 1.0
        
        // Premium-Bonus
        if user.isPremium {
            multiplier *= 1.5
        }
        
        // Event-Boost
        if let activeEvent = getActiveEvent(for: user) {
            multiplier *= activeEvent.boostMultiplier
        }
        
        let coins = Double(steps) * baseRate * multiplier
        user.stompCoins += coins
        
        // Transaktion erstellen
        let transaction = Transaction(
            user: user,
            type: .reward,
            amount: coins,
            currency: .stompCoin,
            description: "Coins generiert durch \(steps) Schritte"
        )
        modelContext.insert(transaction)
        
        try? modelContext.save()
        return coins
    }
    
    // Coin-Konvertierung
    func convertCoins(
        amount: Double,
        from: Currency,
        to: Currency,
        user: User,
        modelContext: ModelContext
    ) -> Bool {
        guard amount > 0 else { return false }
        
        let conversionRate: Double
        switch (from, to) {
        case (.stompCoin, .omCoin):
            conversionRate = 0.1 // 10 StompCoins = 1 OM Coin
            guard user.stompCoins >= amount else { return false }
            user.stompCoins -= amount
            user.omCoins += amount * conversionRate
        case (.omCoin, .stompCoin):
            conversionRate = 10.0 // 1 OM Coin = 10 StompCoins
            guard user.omCoins >= amount else { return false }
            user.omCoins -= amount
            user.stompCoins += amount * conversionRate
        default:
            return false
        }
        
        // Transaktion erstellen
        let transaction = Transaction(
            user: user,
            type: .conversion,
            amount: amount,
            currency: from,
            description: "Konvertierung von \(from.rawValue) zu \(to.rawValue)"
        )
        modelContext.insert(transaction)
        
        try? modelContext.save()
        return true
    }
    
    // Coin-Kauf
    func purchaseCoins(
        amount: Double,
        paymentMethod: PaymentMethod,
        user: User,
        modelContext: ModelContext
    ) -> Bool {
        guard amount > 0 else { return false }
        
        // Hier würde die Zahlungsabwicklung implementiert werden
        // Für jetzt simulieren wir einen erfolgreichen Kauf
        
        user.stompCoins += amount
        
        // Transaktion erstellen
        let transaction = Transaction(
            user: user,
            type: .purchase,
            amount: amount,
            currency: .stompCoin,
            description: "Kauf von \(Int(amount)) StompCoins"
        )
        modelContext.insert(transaction)
        
        try? modelContext.save()
        return true
    }
    
    // Hilfsfunktion für aktive Events
    private func getActiveEvent(for user: User) -> Event? {
        return user.eventHistory.first { $0.event.isActive }
    }
} 

File: StompCoin/StompCoin/Android/app/src/main/java/com/stompcoin/app/data/managers/CoinManager.kt
package com.stompcoin.app.data.managers

import com.stompcoin.app.data.model.Currency
import com.stompcoin.app.data.model.PaymentMethod
import com.stompcoin.app.data.model.Transaction
import com.stompcoin.app.data.model.User
import com.stompcoin.app.data.repository.TransactionRepository
import com.stompcoin.app.data.repository.UserRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CoinManager @Inject constructor(
    private val userRepository: UserRepository,
    private val transactionRepository: TransactionRepository
) {
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    suspend fun convertCoins(
        amount: Double,
        from: Currency,
        to: Currency,
        user: User
    ): Boolean {
        if (amount <= 0) return false

        val conversionRate = when (from to to) {
            Currency.STOMP_COIN to Currency.OM_COIN -> 0.1 // 10 StompCoins = 1 OM Coin
            Currency.OM_COIN to Currency.STOMP_COIN -> 10.0 // 1 OM Coin = 10 StompCoins
            else -> return false
        }

        val hasEnoughCoins = when (from) {
            Currency.STOMP_COIN -> user.stompCoins >= amount
            Currency.OM_COIN -> user.omCoins >= amount
        }

        if (!hasEnoughCoins) return false

        // Coins umrechnen
        when (from) {
            Currency.STOMP_COIN -> {
                user.stompCoins -= amount
                user.omCoins += amount * conversionRate
            }
            Currency.OM_COIN -> {
                user.omCoins -= amount
                user.stompCoins += amount * conversionRate
            }
        }

        // Transaktion erstellen
        val transaction = Transaction(
            userId = user.id,
            type = Transaction.Type.CONVERSION,
            amount = amount,
            currency = from,
            description = "Konvertierung von ${from.name} zu ${to.name}"
        )

        try {
            userRepository.updateUser(user)
            transactionRepository.createTransaction(transaction)
            return true
        } catch (e: Exception) {
            _error.value = e.message
            return false
        }
    }

    suspend fun purchaseCoins(
        amount: Double,
        paymentMethod: PaymentMethod,
        user: User
    ): Boolean {
        if (amount <= 0) return false

        try {
            // Hier würde die Zahlungsabwicklung implementiert werden
            // Für jetzt simulieren wir einen erfolgreichen Kauf
            user.stompCoins += amount

            val transaction = Transaction(
                userId = user.id,
                type = Transaction.Type.PURCHASE,
                amount = amount,
                currency = Currency.STOMP_COIN,
                description = "Kauf von ${amount.toInt()} StompCoins"
            )

            userRepository.updateUser(user)
            transactionRepository.createTransaction(transaction)
            return true
        } catch (e: Exception) {
            _error.value = e.message
            return false
        }
    }
} 